sympy/combinatorics/tests/test_perm_groups.py

```diff
@@ -32,6 +32,13 @@ def test_has():
     assert PermutationGroup(a, b).degree == \
         PermutationGroup(a, b).degree == 6
 
+    # this must be tested by equality since
+    # containment in a set cannot be checked since
+    # g is not hashable
+    g = PermutationGroup(Permutation(0, 2, 1))
+    raises(TypeError, lambda: {g})
+    assert Tuple(1, g).has(g)
```

Comment:
I consider this a bug in PermutationGroup arising from abuse of `__eq__`. It is said that in SymPy `a == b` tests structural equality and that works fine with the `__eq__` method that is defined in `Basic`. Practically any time a subclass overrides the `__eq__` method it leads to problems.

Reply:
This passes all tests in sympy/combinatorics:
```diff
diff --git a/sympy/combinatorics/perm_groups.py b/sympy/combinatorics/perm_groups.py
index c0b91c8..387bc24 100644
--- a/sympy/combinatorics/perm_groups.py
+++ b/sympy/combinatorics/perm_groups.py
@@ -202,7 +202,7 @@ def __contains__(self, i):
     def __len__(self):
         return len(self._generators)
 
-    def __eq__(self, other):
+    def is_equivalent(self, other):
         """Return ``True`` if PermutationGroup generated by elements in the
         group are same i.e they represent the same PermutationGroup.
 
@@ -217,6 +217,8 @@ def __eq__(self, other):
         >>> G.generators == H.generators
         False
         >>> G == H
+        False
+        >>> G.is_equivalent(H)
         True
 
         """
@@ -1786,7 +1788,7 @@ def is_perfect(self):
 
         """
         if self._is_perfect is None:
-            self._is_perfect = self == self.derived_subgroup()
+            self._is_perfect = self.is_equivalent(self.derived_subgroup())
         return self._is_perfect
 
     @property
diff --git a/sympy/combinatorics/tests/test_perm_groups.py b/sympy/combinatorics/tests/test_perm_groups.py
index ff91a33..b496608 100644
--- a/sympy/combinatorics/tests/test_perm_groups.py
+++ b/sympy/combinatorics/tests/test_perm_groups.py
@@ -72,10 +72,12 @@ def test_equality():
     g_1 = PermutationGroup(p_1, p_2)
     g_2 = PermutationGroup(p_3, p_4)
     g_3 = PermutationGroup(p_2, p_1)
+    g_4 = PermutationGroup(p_1, p_2)
 
-    assert g_1 == g_2
+    assert g_1 != g_2
     assert g_1.generators != g_2.generators
-    assert g_1 == g_3
+    assert g_1 != g_3
+    assert g_1 == g_4
 
 
 def test_stabilizer():
```